[chore] Create RFC for Optional config types (#12596)
open-telemetry/opentelemetry-collector: docs/rfcs/optional-config-type.md

# Optional[T] type for use in configuration structs

## Overview

In Go, types are by default set to a "zero-value", a supported value for the
respective type that is semantically equivalent or similar to "empty", but which
is also a valid value for the type. For many config fields, the zero value is
not a valid configuration value and can be taken to mean that the option is
disabled, but in certain cases it can indicate a default value, necessitating a
way to represent the presence of a value without using a valid value for the
type.

Using standard Go without inventing any new types, the two most straightforward
ways to accomplish this are:

1. Using a separate boolean field to indicate whether the field is enabled or
   disabled.
2. Making the type a pointer, which makes a `nil` pointer represent that a
   value is not present, and a valid pointer represent the desired value.

Each of these approaches has deficiencies: Using a separate boolean field
requires the user to set the boolean field to `true` in addition to setting the
actual config option, leading to suboptimal UX. Using a pointer value has a few

> drawbacks:

Reviewer: Mutability is another concern that you mentioned over at https://github.com/open-telemetry/opentelemetry-collector/issues/9478#issuecomment-2414487668 Should we add it here?

Author: I missed that when reviewing the issues related to this PR, but that's a meaningful concern. I've added it.

   a field is optional.
2. The distinction between values that are conventionally pointers (e.g. gRPC
   configs) and optional values is lost.
3. Setting a default value for a pointer field when decoding will set the field
   on the resulting config struct, and additional logic must be done to unset
   the default if the user has not specified a value.
4. The potential for null pointer exceptions is created.

## Optional types

Go does not include any form of Optional type in the standard library, but other
popular languages like Rust and Java do. We could implement something similar in
our config packages that allows us to address the downsides of using pointers
for optional config fields.

## Basic definition

A production-grade implementation will not be discussed or shown here, but the
basic implementation for an Optional type in Go could look something like:

```golang
type Optional[T any] struct {
	hasValue bool
	value    T

	defaultVal T
}

func Some[T any](value T) Optional[T] {
	return Optional[T]{value: value, hasValue: true}
}

func None[T any](value T) Optional[T] {
	return Optional[T]{}
}

> func WithDefault[T any](defaultVal T) Optional[T] {

Reviewer 2: I think `WithDefault` should accept a lambda, `func() T`, not an actual value. The defaults may be runtime sensitive, an example would be the `localhost` vs. `0.0.0.0` that could be controlled by an external mechanism like a feature gate.

Reviewer 2: I had another motivation for function in my prototype https://github.com/open-telemetry/opentelemetry-collector/pull/10260/files#diff-4825c80a1cbc3775a3f3cd1aa7e6d01b6226bc31efaa9ca7b6a1a466b31512f2R36
```
// The reason we pass a function instead of T directly
// is to allow this to work with T being a pointer,
// because we wouldn't want to copy the value of a pointer
// since it might reuse (and override) some shared state.
```

Author: Sorry for the confusion here, the reason I didn't have `WithDefault` accept a function was to simplify the implementation to focus on high-level concepts instead of the design of the API. This is not at all supposed to be a carefully-designed or production-ready implementation, which I try to call out above the code block.

If there's a consideration in the design of `WithDefault` that we should be considering for whether the type is worth adding at all, or if using a value instead of a function makes things less clear, I can adjust the example implementation in this document.

Reviewer 2: fair enough, but the "simplified" implementation is different from the one with lambda in about 10 characters.

	return Optional[T]{defaultVal: defaultVal}
}

func (o Optional[T]) HasValue() bool {
	return o.hasValue
}

func (o Optional[T]) Value() T {
    return o.value
}

func (o Optional[T]) Default() T {
    return o.defaultVal
}
```

## Use cases

Optional types can fulfill the following use cases we have when decoding config.

> ### Representing optional config fields

Reviewer: Could we list other use cases? e.g. TLSConfig

Author: Sure thing, I added another example for how this would impact `confighttp.ServerConfig`. It doesn't look substantially different from how it's used in the OTLP receiver, but it should help illustrate how this might look.

            }),
		},
	}
}
```

### Proper unmarshaling of empty values when a default is set

Currently, the OTLP receiver requires a workaround to make enabling each
protocol optional while providing defaults if a key for the protocol has been
set:

```golang
type Protocols struct {
	GRPC *configgrpc.ServerConfig `mapstructure:"grpc"`
	HTTP *HTTPConfig              `mapstructure:"http"`
}

// Config defines configuration for OTLP receiver.
type Config struct {
	// Protocols is the configuration for the supported protocols, currently gRPC and HTTP (Proto and JSON).
	Protocols `mapstructure:"protocols"`
}

func createDefaultConfig() component.Config {
	return &Config{
		Protocols: Protocols{
			GRPC: configgrpc.NewDefaultServerConfig(),
			HTTP: &HTTPConfig{
                // ...
			},
		},
	}
}

func (cfg *Config) Unmarshal(conf *confmap.Conf) error {
	// first load the config normally
	err := conf.Unmarshal(cfg)
	if err != nil {
		return err
	}

    // gRPC will be enabled if this line is not run
	if !conf.IsSet(protoGRPC) {
        cfg.GRPC = nil
	}

     // HTTP will be enabled if this line is not run
	if !conf.IsSet(protoHTTP) {
		cfg.HTTP = nil
	}

	return nil
}
```

> With an Optional type, the checks in `Unmarshal` become unnecessary, and it's

Reviewer 2: having `Unmarshal` by itself is not a problem, the problem is that it cannot be made type-safe, since `conf.IsSet(protoGRPC)` checks against a string with field name, disconnected from the actual field name.

Author: Fully agreed, that's why I call out the checks in Unmarshal here as opposed to the function itself. Granted, in this case specifically, this implementation of `Unmarshal` could be removed since it only handles unsetting default values.

possible the entire `Unmarshal` function may no longer be needed. Instead, when

> the the config is unmarshaled, no value would be put into the default Optional

Reviewer: ```suggestion
the config is unmarshaled, no value would be put into the default Optional
```

values and `HasValue` would return false when using this config object.

This situation is something of an edge case with our current unmarshaling
facilities and while not common, could be a rough edge for users looking to
implement similar config structures.

## Disadvantages of an Optional type

There are two noteworthy disadvantages of introducing an Optional type:

1. Since the type isn't standard, external packages working with config may

>    require additional adaptations to work with our config structs.

Reviewer 2: Can we give examples of external packages working with config? E.g. validation libs come to mind, but they might still work, for them Optional is just an extra nesting container.

Reviewer 2: ok, but how do you want to move the discussion forward? Just take a vote? Or do more cost/benefit analysis, and potentially block this on unresolved #10694?

Author: We'll essentially need a vote to get this merged anyway per the Collector's [RFC process](https://github.com/open-telemetry/opentelemetry-collector/blob/main/docs/rfcs/README.md#merge-rules), so I think that's what it will come down to.

I think this is far more straightforward than https://github.com/open-telemetry/opentelemetry-collector/pull/10694 anyway, so I would prefer to not block it on that. I'm just calling it out here as a risk.

Reviewer: I think there is no specific concern on this thread, and the JSON schema libraries support custom types so I am resolving this comment

> 2. It will still be possible to use pointer types for optional config fields,

Reviewer 2: I don't think this is a disadvantage of the Optional type specifically. There can be a feature gate added that will run a validator that prohibits (or warns about) pointer fields, to ensure consistency.

Reviewer: Agreed, I think we have the same issue with, say, `configopaque`, and it's so far fine (Evan can keep me honest here wrt e.g. serializing)

Author: That's a good callout, I'll remove this point.

## Overview

In Go, types are by default set to a "zero-value", a supported value for the
respective type that is semantically equivalent or similar to "empty", but which
is also a valid value for the type. For many config fields, the zero value is
not a valid configuration value and can be taken to mean that the option is
disabled, but in certain cases it can indicate a default value, necessitating a
way to represent the presence of a value without using a valid value for the
type.

Using standard Go without inventing any new types, the two most straightforward
ways to accomplish this are:

1. Using a separate boolean field to indicate whether the field is enabled or
   disabled.
2. Making the type a pointer, which makes a `nil` pointer represent that a
   value is not present, and a valid pointer represent the desired value.

Each of these approaches has deficiencies: Using a separate boolean field
requires the user to set the boolean field to `true` in addition to setting the
actual config option, leading to suboptimal UX. Using a pointer value has a few
drawbacks:

1. It may not be immediately obvious to a new user that a pointer type indicates
   a field is optional.

> 2. The distinction between values that are conventionally pointers (e.g. gRPC
>    configs) and optional values is lost.
> 3. Setting a default value for a pointer field when decoding will set the field
>    on the resulting config struct, and additional logic must be done to unset
>    the default if the user has not specified a value.
> 4. The potential for null pointer exceptions is created.
> 5. Config structs are generally intended to be immutable and may be passed
>    around a lot, which makes the mutability property of pointer fields
>    an undesirable property.

Reviewer 3: Drawback numbers 2 and 5 appear contradictory. Is there a reason why pointers are "conventionally" pointers, aside from mutability?
```suggestion
2. While they can be copied cheaply, pointers convey a mutable reference making
   them unsafe to share.
3. Setting a default value for a pointer field when decoding will set the field
   on the resulting config struct, and additional logic must be done to unset
   the default if the user has not specified a value.
4. The potential for null pointer exceptions is created.
```

Author: This may not really apply to config structs, but outside mutability, another reason to use a pointer type for a field in Go is so you don't have to pass around a large struct by value.

The specific point I was making with item 2 is that some types in popular libraries are pointers, e.g. `*zap.Logger`, `*grpc.Server`, etc. If pointers are used to indicate optional values, if you have a field that is a `*zap.Logger` type, it's difficult to tell if it's intentionally optional, or if that type is almost always used as a pointer type (`zap.Logger` as a non-pointer is rare from what I've seen). Obviously `*zap.Logger` isn't a perfect example because nobody is setting a logger in YAML, but hopefully my point still makes sense.

Reviewer: @jmacd does Evan's comment solve your question?

our config packages that allows us to address the downsides of using pointers
for optional config fields.

## Basic definition

A production-grade implementation will not be discussed or shown here, but the
basic implementation for an Optional type in Go could look something like:

```golang
type Optional[T any] struct {
	hasValue bool
	value    T

	defaultVal T
}

func Some[T any](value T) Optional[T] {
	return Optional[T]{value: value, hasValue: true}
}

func None[T any](value T) Optional[T] {
	return Optional[T]{}
}

func WithDefault[T any](defaultVal T) Optional[T] {
	return Optional[T]{defaultVal: defaultVal}
}

func (o Optional[T]) HasValue() bool {
	return o.hasValue
}

func (o Optional[T]) Value() T {
    return o.value
}

func (o Optional[T]) Default() T {
    return o.defaultVal
}
```

## Use cases

Optional types can fulfill the following use cases we have when decoding config.

### Representing optional config fields

To use the optional type to mark a config field as optional, the type can simply
be used as a type parameter to `Optional[T]`. The following config struct shows
how this may look, both in definition and in usage:

> ```golang

Reviewer 4: More than golang code, would be good to show how users in yaml will use it.

Reviewer 2: there's no expected impact on YAML

Reviewer: @bogdandrutu does Yuri's comment solve your question?

Reviewer: Clarified inline on 5d5f6527f1d4444c657d30c19f3055e781830913

	Protocols `mapstructure:"protocols"`
}

func createDefaultConfig() component.Config {
	return &Config{
		Protocols: Protocols{
			GRPC: configgrpc.NewDefaultServerConfig(),
			HTTP: &HTTPConfig{
                // ...
			},
		},
	}
}

func (cfg *Config) Unmarshal(conf *confmap.Conf) error {
	// first load the config normally
	err := conf.Unmarshal(cfg)
	if err != nil {
		return err
	}

    // gRPC will be enabled if this line is not run
	if !conf.IsSet(protoGRPC) {
        cfg.GRPC = nil
	}

     // HTTP will be enabled if this line is not run
	if !conf.IsSet(protoHTTP) {
		cfg.HTTP = nil
	}

	return nil
}
```

With an Optional type, the checks in `Unmarshal` become unnecessary, and it's
possible the entire `Unmarshal` function may no longer be needed. Instead, when
the config is unmarshaled, no value would be put into the default Optional
values and `HasValue` would return false when using this config object.

This situation is something of an edge case with our current unmarshaling
facilities and while not common, could be a rough edge for users looking to
implement similar config structures.

## Disadvantages of an Optional type

> There are two noteworthy disadvantages of introducing an Optional type:

Reviewer: ```suggestion
There is one noteworthy disadvantage of introducing an Optional type:
```

> 1. Since the type isn't standard, external packages working with config may

Reviewer 5: What is the intended implementation plan for Optional config types? In the above example, it says:
>This situation is something of an edge case

So is the intention to only modify fields who meet some required use case? Or is it to modify all fields that are currently pointers?

Changing the field type is a breaking API change. Is there a different approach depending on the stability level of a component? Or is the expectation that anyone importing and using a config struct from an otel component will have to make any required changes given the new struct field types?

Reviewer: @kristinapathak Do you have examples of components/modules that are 1.x and need to use this optional type? We are blocking `confighttp`/`configgrpc` on this so that we can apply this pattern on these modules.

The intent is to apply this new type everywhere, go through an (API) breaking change, and ensure this type is used everywhere going forward.

Reviewer: @kristinapathak does this solve your question?

Reviewer 5: yes, thank you
